WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java

```diff
@@ -71,154 +78,298 @@ public void fetchBlogList(Callback callback) {
             mErrorMsgId = org.wordpress.android.R.string.invalid_site_url_message;
             return null;
         }
-        URI uri = URI.create(baseUrl);
+
+        AppLog.i(T.NUX, "Trying system.listMethods on the following URL: " + url);
+        URI uri = URI.create(url);
         XMLRPCClientInterface client = XMLRPCFactory.instantiate(uri, mHttpUsername, mHttpPassword);
-        try {
-            client.call(Method.LIST_METHODS);
-            xmlRpcUrl = baseUrl;
-            return xmlRpcUrl;
-        } catch (XMLRPCException e) {
-            AppLog.i(T.NUX, "system.listMethods failed on: " + baseUrl);
-            if (isHTTPAuthErrorMessage(e)) {
-                return null;
+        return client.call(Method.LIST_METHODS);
+    }
+
+    private boolean validateListMethodsResponse(Object[] availableMethods) {
+        if (availableMethods == null) {
+            AppLog.e(T.NUX, "The response of system.listMethods was empty!");
+            return false;
+        }
+        // validate xmlrpc methods
+        String[] requiredMethods =  { "wp.getUsersBlogs", "wp.getPage", "wp.getCommentStatusList", "wp.newComment",
+                "wp.editComment", "wp.deleteComment", "wp.getComments",	"wp.getComment",
+                "wp.getOptions", "wp.uploadFile", "wp.newCategory",
+                "wp.getTags", "wp.getCategories", "wp.editPage", "wp.deletePage",
+                "wp.newPage", "wp.getPages" };
+
+        for (String currentRequiredMethod: requiredMethods) {
+            boolean match = false;
+            for (Object currentAvailableMethod: availableMethods) {
+                if ((currentAvailableMethod).equals(currentRequiredMethod)) {
+                    match = true;
+                    continue;
+                }
             }
-        } catch (SSLHandshakeException e) {
-            if (!WPUrlUtils.isWordPressCom(baseUrl)) {
-                mErroneousSslCertificate = true;
+
+            if (!match) {
+                AppLog.e(T.NUX, "The following XML-RPC method: " + currentRequiredMethod + " is missing on the server.");
+                return false;
             }
-            AppLog.w(T.NUX, "SSLHandshakeException failed. Erroneous SSL certificate detected.");
-            return null;
-        } catch (SSLPeerUnverifiedException e) {
-            if (!WPUrlUtils.isWordPressCom(baseUrl)) {
-                mErroneousSslCertificate = true;
+        }
+        return true;
+    }
+
+    // Append "xmlrpc.php" if missing in the URL
+    private String appendXMLRPCPath(String url) throws IllegalArgumentException {
+        if (!URLUtil.isValidUrl(url)) {
+            throw new IllegalArgumentException("Input URL " + url + " is not valid!");
+        }
+
+        if (!url.contains("xmlrpc.php")) { // Do not use 'ends' here! Some hosting wants parameters passed to baseURL/xmlrpc-php?my-authcode=XXX
+            if (url.substring(url.length() - 1, url.length()).equals("/")) {
```

Comment:
Very minor nitpick: Using `charAt` might makes this a bit more readable:

`if (url.charAt(url.length() - 1) == '/')`